### PR TITLE
Initial Journal Implementation

### PR DESCRIPTION
**Pull request by frrist**
### Goals
Make progress towards addressing #3269.

### Implementation
- Define a new package `journal/`.
- A journal is a single file that all topics and their associated operations and metadata are written to. And example of this file is shown by [this gist.](https://gist.github.com/frrist/b8014e16d117c854d3772bbf229805c8) The journal file will be kept in `$REPODIR/filecoin_journal.json.`
- Define a `journal.Writer` type, services that wish to keep a journal may accept a `journal.Writer` and use it for recording events. This writer has been added to the message outbox.

### CODE REVIEW

File: journal/journal.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	// Record records an operation and its metadata to a Journal accepting variadic key-value
+	// pairs.
+	Record(operation string, meta ...interface{})
+}
+
+// Builder defines a method for creating Journals with a topic.
+type Builder func(topic string) Journal
+
+// NewZapJournalBuilder returns a Builder backed by a zap logger.
+func NewZapJournalBuilder(filepath string) (Builder, error) {
```
Comment (anorth) [inline journal/journal.go]: I think the topic string should be provided from the outside rather than determined within a component. So some naming things:
- NewZapJournalBuilder -> New
- JournalBuilder -> Journal
- Journal -> Writer
- Record() -> Write()

Rather than a func, use an interface for the ~Builder~Journal.

In the node, we'll then have:
```
j := journal.New(path)
syncer := chain.NewSyncer(..., j.Topic("chain"))
```
and in a component
```
  self.journal.Write("event", "key", value")
```

File: journal/journal.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// ZapJournal implemented the Journal interface.
+type ZapJournal struct {
+	logger *zap.SugaredLogger
+}
+
+// Record records an operation and its metadata to a Journal accepting variadic key-value
+// pairs.
+func (zj *ZapJournal) Record(operation string, kv ...interface{}) {
+	zj.logger.Infow(operation, kv...)
+}
```
Comment (anorth) [inline journal/journal.go]: Move the Zap impl to `journal/zap.go` and implement an in-memory journal here (for tests).

File: message/pool.go [hunk 3]
```diff
@@ -34,6 +35,8 @@ type Pool struct {
 	validator     PoolValidator
 	pending       map[cid.Cid]*timedmessage // all pending messages
 	addressNonces map[addressNonce]bool     // set of address nonce pairs used to efficiently validate duplicate nonces
+
+	journal journal.Journal
```
Comment (anorth) [inline message/pool.go]: For now, don't journal this. It's massively higher traffic than the outbox and the need is less clear. I think we should do the command API before this, but in any case let's defer that discussion to new PRs.

File: node/builder.go [hunk 4]
```diff
@@ -163,6 +165,15 @@ func (nc *Builder) build(ctx context.Context) (*Node, error) {
 		return nil, err
 	}
 
+	repoPath, err := nc.Repo.Path()
+	if err != nil {
+		return nil, err
+	}
+
+	jBuilder, err := journal.NewZapJournalBuilder(fmt.Sprintf("%s/%s", repoPath, "journal.json"))
```
Comment (anorth) [inline node/builder.go]: The repo should be in control of its internals. Ask the repo to construct the path.

File: commands/daemon.go [hunk 5]
```diff
@@ -96,6 +97,12 @@ func daemonRun(req *cmds.Request, re cmds.ResponseEmitter) error {
 	opts = append(opts, node.BlockTime(blockTime))
 	opts = append(opts, node.ClockConfigOption(clock.NewSystemClock()))
 
+	journal, err := journal.NewZapJournal(rep.JournalPath())
+	if err != nil {
+		return err
+	}
+	opts = append(opts, node.JournalConfigOption(journal))
+
```
Comment (frrist) [inline commands/daemon.go]: Only create a zap backed journal when something runs the daemon via the `daemon` command.

File: journal/journal.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+// Journal defines an interface for creating Journals with a topic.
+type Journal interface {
+	// Topic returns a Writer that records events for a topic.
+	Topic(topic string) Writer
+}
+
+// NewInMemoryJournal returns a journal backed by an in-memory map.
+func NewInMemoryJournal(clk clock.Clock) Journal {
```
Comment (frrist) [inline journal/journal.go]: Thinking it might make more sense for this to go in a testing package as that's all I image we'll use it for, thoughts?

File: repo/fsrepo.go [hunk 7]
```diff
@@ -557,6 +557,11 @@ func (r *FSRepo) Path() (string, error) {
 	return r.path, nil
 }
 
+// JournalPath returns the path the journal is at.
+func (r *FSRepo) JournalPath() string {
+	return fmt.Sprintf("%s/filecoin_journal.json", r.path)
+}
+
```
Comment (frrist) [inline repo/fsrepo.go]: This should probably be configurable, but doing so now feels out of scope. I propose that if we want a configurable path we do it in a follow on.

File: journal/journal.go (hunk 6 above)
Comment (frrist) [reply journal/journal.go]: I would also be open to a `NoopJournal`, getting this to have the same behaviour as the `ZapJournal` isn't simple.

File: journal/journal.go [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+// Write records an operation and its metadata to a Journal accepting variadic key-value
+// pairs.
+func (mw *MemoryWriter) Write(event string, kvs ...interface{}) {
+	var ent entry
+	ent.time = mw.clock.Now()
+	ent.event = event
+	extracted, err := extractKeyValues(kvs...)
+	if err != nil {
+		panic(err)
+	}
```
Comment (frrist) [inline journal/journal.go]: The zapjournal writes a error message to a configurable location, currently this is set to stderr: https://github.com/filecoin-project/go-filecoin/pull/3485/files#diff-c65947145146f8d0a4bae969afded6abR18

I _really_ don't want the `Write` interface to return an error, we could do something similar to zap here or just ignore the errors, thoughts?

File: journal/zap.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+// NewZapJournal returns a Builder backed by a zap logger.
+func NewZapJournal(filepath string) (Journal, error) {
+	zapCfg := zap.NewProductionConfig()
+	zapCfg.Encoding = "json"
+	zapCfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+	zapCfg.EncoderConfig.LevelKey = ""
+	zapCfg.EncoderConfig.CallerKey = ""
+	zapCfg.EncoderConfig.MessageKey = "operation"
+	zapCfg.EncoderConfig.NameKey = "topic"
```
Comment (anorth) [inline journal/zap.go]: To avoid name collisions, I recommend a prefix, e.g. `_topic`. And "event" rather than "operation"?

File: message/outbox.go [hunk 10]
```diff
@@ -78,6 +82,7 @@ func (ob *Outbox) Send(ctx context.Context, from, to address.Address, value type
 		if err != nil {
 			msgSendErrCt.Inc(ctx, 1)
 		}
+		ob.journal.Write("send", "to", to.String(), "from", from.String(), "method", method, "error", err)
```
Comment (anorth) [inline message/outbox.go]: Value, gas price, gas limit, params - all of them

File: message/outbox_test.go [hunk 11]
```diff
@@ -29,7 +29,8 @@ func TestOutbox(t *testing.T) {
 		provider := message.NewFakeProvider(t)
 		bcast := true
 
-		ob := message.NewOutbox(w, message.FakeValidator{RejectMessages: true}, queue, publisher, message.NullPolicy{}, provider, provider)
+		ob := message.NewOutbox(w, message.FakeValidator{RejectMessages: true}, queue, publisher,
+			message.NullPolicy{}, provider, provider, newOutboxTestJournal())
```
Comment (anorth) [inline message/outbox_test.go]: Don't depend across test files like this. Move the function here, and just use a single variable in handler_integration_test.

File: repo/fsrepo.go (hunk 7 above)
Comment (anorth) [reply repo/fsrepo.go]: Non-configurable is fine. But drop the redundant `filecoin` in the file name.

**Comment by frrist** [comment]
Follow on issue to instrument elsewhere: https://github.com/filecoin-project/go-filecoin/issues/3556
